### DISCUSSION TITLE
Harmonized build/folder structure for different OS

### DISCUSSION THREAD

**Discussion by nefarius2001** [jamulussoftware/jamulus#2153] [Ideas]
There are autobuild- and deploy- scripts in:
jamulus/mac for MacOS
jamulus/linux for Linux
jamulus/distribute_android for Android (intentional, to not mix code and build-scripts)
jamulus/windows for Windows (in progress)

I am confused by the directory jamulus/distributions. 
- Do we still need debian scripts there, when there are others labeled linux?
- Are these files promoted in any instructions?

I propose to have all build- and deploy- related stuff in distributions like:
jamulus/distributions/macos
jamulus/distributions/linux 
jamulus/distributions/windows 
jamulus/distributions/android
jamulus/distributions/raspi

**Comment by pljones** [comment]
Save me trying to work it out, why is `raspi` not a Linux flavour?

**Comment by ann0see** [comment]
>  why is raspi not a Linux flavour

Same thought here: Couldn't we move it to the installscripts folder?

**Comment by nefarius2001** [comment]
I see two cathegories: autobuild (create a file that can be distributed) and installscripts (setting up a system). The raspi-file is clearly an installscript to me.

Me personally, I am confused by the folder name "distributions".
Here is an updated proposal:

```
jamulus/autobuild/macos/
jamulus/autobuild/linux/
jamulus/autobuild/linux_flatpack/  (io.jamulus.Jamulus.*)
jamulus/autobuild/linux_debian_package/
jamulus/autobuild/windows/
jamulus/autobuild/android/
jamulus/installscripts/    (install4*.sh, raspijamulus.sh)
```

@ann0see which of the  resouces/folders are used by debuild?

**Comment by ann0see** [comment]
> Me personally, I am confused by the folder name "distributions".

Me too.

As far as I know, the .deb generation script copies the debian folder from distributions/debian to the repo root. This means that the debian scripts have to be modified. I think @mirabilos might have a deeper insight here. 

@niebert created the install4 scripts and (as far as I remember), he wanted to maintain them himself somewhere else. Maybe we could put them on the documentation repo?

your proposal looks good, but why don't we use:
```
jamulus/mac/autobuild
jamulus/linux/autobuild
jamulus/linux/autobuild/flatpak/  (io.jamulus.Jamulus.*)
jamulus/linux/autobuild/debian/
jamulus/linux/installscripts/    (install4*.sh, raspijamulus.sh)
jamulus/windows/autobuild/
jamulus/android/autobuild/
```

This would be more consistent to the current structure.

Also there's another issue: Should we put the translations of the installer to /src/res/translation/WinInstaller ?

**Comment by nefarius2001** [comment]
I like to have all code clearly (top-level) separated from scripting-stuff (installscripts and autobuild). But I see the beauty in ann0sees proposal, too. Might be more intuitive, even. Given, that there are not other subfolders (where translation and ASIOSDK2 for windows are the only ones currently)

Other's opinions on this?

**Comment by corrados** [comment]
I prefer ann0see's proposal.

**Comment by mirabilos** [comment]
ann0see dixit:

>from distributions/debian to the repo root. This means that the debian
>scripts have to be modified. I think @mirabilos might have a deeper
>insight here.

The Debian package of Jamulus does not use the distributions/ directory
of Jamulus, other than what Jamulus.pro does (install the desktop file
and the icon).

So I have no stake in this.

>As far as I know, the .deb generation script copies the debian folder

That’s just that, a generation script for a binary .deb package that
can be installed on dpkg-using distributions. That’s not a Debian pak‐
kage ;-) we always build from source on the build dæmons, using Debian-
specific packaging.

Indeed, if there were a top-level debian/ subdirectory, the method
used to unpack the .tar.gz/xz would completely remove it, because a
top-level debian/ directory is reserved for the packaging.

bye,
//mirabilos
-- 
22:20⎜<asarch> The crazy that persists in his craziness becomes a master
22:21⎜<asarch> And the distance between the craziness and geniality is
only measured by the success 18:35⎜<asarch> "Psychotics are consistently
inconsistent. The essence of sanity is to be inconsistently inconsistent

**Comment by pljones** [comment]
Current the linux, mac and windows have a mixture of stuff.

I'd prefer what @nefarius2001 said and reorganise so that:
- `src/` for _all_ compilable source code with `src/<platform>` for `<platform>`-specific bits, possibly with `src/<platform>/<sub-platform>` if needed (e.g. difference Android versions or something)
- `dependency/<component>` or `dependency/<platform>/<component>` for stuff that compiling the code needs to have built before compile and link will complete successfully
- `autobuild/` for all the scripts necessary for automated build with `autobuild/<platform>` for platform-specific tools and, again, `autobuild/<platform>/<sub-platform>` if necessary

That way, if I want Jamulus source, I go to the Jamulus source directory.  I don't need to think "is this going to have different versions on different platforms?" before I start looking for something.  As in `grep -r "something" src/`...

**Comment by ann0see** [comment]
That would be possible too but require some more restructuring/fixing paths in the source, ...

**Comment by nefarius2001** [comment]
I’dconcentrate on clearing the build stuff first, then in a second step the dependencies/ resources

**Comment by nefarius2001** [comment]
So it’s 2:2 now, regarding the proposals.
Is a top Level „Autobuild“ folder acceptable,  too? @corrados @ann0see

**Comment by ann0see** [comment]
For me: acceptable but not "beautiful".

**Comment by gilgongo** [comment]
This feels like it needs some more discussion.

**Comment by ann0see** [comment]
Related: https://github.com/jamulussoftware/jamulus/pull/2576

The repo will look quite different in future...

**Comment by niebert** [comment]
Created the scripts mentioned above at a time where automated step by step instructions where availble on the Jamulus homepage which was developed and provided by Volker https://github.com/corrados.  Feel free to remove all my install scripts due to the fact other installation methods are available especially on Linux systems including Raspberry Pi.  Anyway, would recommend to do proper acknowledgement of Volkers  tremendously good work on Jamulus  e.g. in the README.md It was mainly a one-man-show of software development for Jamuluts including the theoretical foundation of Jamulus  (2015) Case Study: Performing Band Rehearsals on the Internet WithJamulus (Date: 2015/02/24) URL: https://jamulus.io/PerformingBandRehearsalsontheInternetWithJamulus.pdf - (accessed 2020/11/08). Thank you,

**Reply by ann0see** [reply]
Yes, concerning the acknowledgements we can surely put that in the README file. @gilgongo?

Concerning the install scripts, they are on a separate repo now – which I gave you access to, I think https://github.com/jamulussoftware/installscripts

Maybe you didn't accept the invite? Just sent another one.

**Comment by niebert** [comment]
No worries, the isntall script were a minor contribution to the repository, feel free to remove anything, that is not of value for the Jamulus respo.
